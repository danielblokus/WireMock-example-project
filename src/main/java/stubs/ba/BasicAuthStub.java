package stubs.ba;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static utils.HttpStatusCodes.SUCCESS;
import static utils.WebServicePaths.BA_ENDPOINT_PATH;

public class BasicAuthStub {

    private final String bodyResponse = "Hello BA!";
    private static final String VALID_USER_NAME = "admin@danielblokus.github.io";
    private static final String VALID_PASSWORD = "12345678";

    public void stubResponse() {
        stubFor(get(BA_ENDPOINT_PATH)
                .withBasicAuth(VALID_USER_NAME, VALID_PASSWORD)
                .willReturn(aResponse()
                        .withBody(bodyResponse)
                        .withStatus(SUCCESS)));
    }
}