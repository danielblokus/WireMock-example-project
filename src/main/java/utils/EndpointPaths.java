package utils;

public final class EndpointPaths {

	public static final String EXAMPLE_ENDPOINT_PATH = "/example";
	public static final String INVALID_ENDPOINT_PATH = "/something/else";

	private EndpointPaths() {

	}
}
