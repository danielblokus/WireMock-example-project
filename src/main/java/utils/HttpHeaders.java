package utils;

public final class HttpHeaders {

    public static final String CONTENT_TYPE_HEADER_KEY = "Content-Type";

    public static final String JSON_HEADER_VALUE = "application/json";

    private HttpHeaders() {}
}