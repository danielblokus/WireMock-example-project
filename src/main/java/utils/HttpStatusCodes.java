package utils;

public final class HttpStatusCodes {

	public static final int SUCCESS = 200;
	public static final int NOT_FOUND = 404;

	private HttpStatusCodes() {

	}
}