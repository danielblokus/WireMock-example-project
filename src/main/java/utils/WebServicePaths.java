package utils;

public final class WebServicePaths {

    public static final String EXAMPLE_ENDPOINT_PATH = "/example";
    public static final String INVALID_ENDPOINT_PATH = "/something/else";
    public static final String BA_ENDPOINT_PATH = "/basic/auth";

    private WebServicePaths() {

    }
}