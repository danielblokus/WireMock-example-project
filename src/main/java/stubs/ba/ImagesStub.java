package stubs.ba;

import utils.HttpHeaders;
import utils.HttpStatusCodes;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.get;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static utils.WebServicePaths.IMAGES_ENDPOINT_PATH;

public class ImagesStub {

    private static final String JSON_RESPONSE_PATH = "stubs/ba/imagesResponse.json";
    private static final String VALID_USER_NAME = "admin@danielblokus.github.io";
    private static final String VALID_PASSWORD = "12345678";

    public void stubResponse() {
        stubFor(get(IMAGES_ENDPOINT_PATH)
                .withBasicAuth(VALID_USER_NAME, VALID_PASSWORD)
                .willReturn(aResponse()
                        .withStatus(HttpStatusCodes.SUCCESS)
                        .withHeader(HttpHeaders.CONTENT_TYPE_HEADER_KEY, HttpHeaders.JSON_HEADER_VALUE)
                        .withHeader("User-Id", generateRandomUUID())
                        .withBodyFile(JSON_RESPONSE_PATH)));
    }

    private String generateRandomUUID() {
        return UUID.randomUUID().toString();
    }
}