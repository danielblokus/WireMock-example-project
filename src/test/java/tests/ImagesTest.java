package tests;

import com.github.tomakehurst.wiremock.client.WireMock;
import com.jayway.jsonpath.JsonPath;
import io.restassured.response.Response;
import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Test;
import stubs.ba.ImagesStub;
import utils.HttpStatusCodes;
import utils.WebServicePaths;

import java.util.List;

import static com.github.tomakehurst.wiremock.client.WireMock.getRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
import static io.restassured.RestAssured.given;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class ImagesTest extends BasicTest {

    private static final String HTTPS_REGEX = "^(https)://.*$";
    private static final String ORIGINAL_URL_JSON_PATH = "$..originalUrl";

    private final ImagesStub imagesStub = new ImagesStub();

    @Before
    public void setUpEnvironment() {
        imagesStub.stubResponse();
    }

    @Test
    public void checkBodyResponseNotEmpty() {
        given()
        .when()
            .get(WebServicePaths.IMAGES_ENDPOINT_PATH)
        .then()
            .assertThat()
                .body("results", hasSize(greaterThan(0)));

        WireMock.verify(getRequestedFor(urlPathEqualTo(WebServicePaths.IMAGES_ENDPOINT_PATH)));
    }

    @Test
    public void checkStatusCode() {
        given()
        .when()
            .get(WebServicePaths.IMAGES_ENDPOINT_PATH)
        .then()
            .assertThat()
                .statusCode(HttpStatusCodes.SUCCESS);
    }


    @Test
    public void checkOriginalImagesUrlsHaveHttpsScheme() {
        getListOfOriginalUrls().forEach(x -> assertThat(x.matches(HTTPS_REGEX), is(true)));
    }

    private List<String> getListOfOriginalUrls() {
        Response response = given()
            .when()
                .get(WebServicePaths.IMAGES_ENDPOINT_PATH)
            .then()
                .extract().response();
        return JsonPath.read(response.asString(), ORIGINAL_URL_JSON_PATH);
    }

    @Test
    public void checkEntireResponse() {
        Response response = given()
                .get(WebServicePaths.IMAGES_ENDPOINT_PATH);
        String responseAsString = response.headers().toString() + "\n" + response.body().asString();
        responseAsString = responseAsString.replaceAll("User-Id=.*", "###GENERATED-USER-ID###");
        responseAsString = responseAsString.replaceAll("Matched-Stub-Id=.*", "###GENERATED-STUB-ID###");

        Approvals.verify(responseAsString);
    }
}