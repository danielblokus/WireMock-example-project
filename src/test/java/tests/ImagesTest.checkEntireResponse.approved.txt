Content-Type=application/json
###GENERATED-USER-ID###
###GENERATED-STUB-ID###
Content-Encoding=gzip
Vary=Accept-Encoding, User-Agent
Transfer-Encoding=chunked
Server=Jetty(9.2.26.v20180806)
[
  {
    "id": 1,
    "title": "",
    "originalUrl": "https://fakeimg.pl/180x180/ff0000,128/333333,255/?text=1&font=roboto",
    "thumbnailUrl": "https://fakeimg.pl/60x60/ff0000,128/333333,255/?text=1&font=roboto"
  },
  {
    "id": 2,
    "title": "",
    "originalUrl": "https://fakeimg.pl/180x180/ff0000,128/333333,255/?text=2&font=roboto",
    "thumbnailUrl": "https://fakeimg.pl/60x60/ff0000,128/333333,255/?text=2&font=roboto"
  },
  {
    "id": 3,
    "title": "",
    "originalUrl": "https://fakeimg.pl/180x180/ff0000,128/333333,255/?text=3&font=roboto",
    "thumbnailUrl": "https://fakeimg.pl/60x60/ff0000,128/333333,255/?text=3&font=roboto"
  },
  {
    "id": 4,
    "title": "",
    "originalUrl": "https://fakeimg.pl/180x180/ff0000,128/333333,255/?text=4&font=roboto",
    "thumbnailUrl": "https://fakeimg.pl/60x60/ff0000,128/333333,255/?text=4&font=roboto"
  },
  {
    "id": 5,
    "title": "",
    "originalUrl": "https://fakeimg.pl/180x180/ff0000,128/333333,255/?text=5&font=roboto",
    "thumbnailUrl": "https://fakeimg.pl/60x60/ff0000,128/333333,255/?text=5&font=roboto"
  }
]